test/config_tests.ml

```diff
@@ -0,0 +1,124 @@
+let pmsg =
+  Alcotest.testable (fun ppf (`Msg s) -> Fmt.pf ppf "Error @[<v>(%s)@]" s)
+    (=)
+
+let result_p_err (type x) ok =
+  (* like Alcotest.result, but pretty-prints the Error `Msg and
+     fails the test when [Error _] is encountered.
+     On success, [ok]'s [pp] and [equal]
+     function are used to pretty-print and determine test result. *)
+  let module Ok = (val ok : Alcotest.TESTABLE with type t = x) in
+  let pp = (fun ppf -> function
+      | Ok v -> Fmt.pf ppf "Ok @[<v>(%a)@]" Ok.pp v
+      | Error `Msg v -> Fmt.pf ppf "Error @[<v>(%s)@]" v) in
+  Alcotest.testable pp
+    (fun a b -> match a,b with
+       | Ok a, Ok b -> Ok.equal a b
+       | _ -> false)
+
+let conf_map = Alcotest.testable
+    Openvpn.Config.pp Openvpn.Config.(equal eq)
+
+let parse_noextern conf =
+  Openvpn.Config.parse_client ~string_of_file:(fun path ->
+      Rresult.R.error_msgf
+        "this test suite does not read external files, \
+         but a config asked for: %S" path) conf
+
+
+let ok_minimal_client () =
+  (* verify that we can parse a minimal good config. *)
+  let basic =
+    {|tls-client
+    auth-user-pass [inline]
+    <auth-user-pass>
+testuser
+testpass
+</auth-user-pass>
+remote 10.0.0.1|} in
+  Alcotest.(check (result pass pmsg)) "basic conf works"
+    (Ok Openvpn.Config.empty)
+    (parse_noextern basic)
+
+let auth_user_pass_trailing_whitespace () =
+  (* Seems to me that the OpenVPN upstream accepts this and we don't.
+     It should also be tested if the upstream version strips prefixed/trailing
+     whitespace from the user/pass/end-block lines. TODO.
+  *)
+  let common =
+    "client\n"
+    ^ "remote 127.0.0.1\n"
+    ^ "auth-user-pass [inline]\n"
+    ^ "<auth-user-pass>\n"
+    ^ "testuser\n"
+    ^ "testpass\n" in
```

Comment:
I find the multi-line string literal syntax (`{| ... |}`) easier to read than these concatenated strings..

Reply:
I had problems with auto-indentation there since the semantics are different; leading whitespace in `{| |}` is included in the string, and this test explicitly deals with whitespace, so I wanted to be sure of what ended up in this test even after eventual reformatting. Generally I agree :)